NPE on instantiation when GroupID missing

Exception appeared in io.confluent.parallelconsumer.ParallelEoSStreamProcessor.checkAutoCommitIsDisabled.
Looks like `(ConsumerCoordinator)coordinatorField.get(consumer)` returns a null;

The code of a creation of ParallelStreamProcessor:

```
@Slf4j
@Component
public class ParallelConsumer {
    String inputTopic = "test-topic";

    public ParallelConsumer() {
        consume();
    }

    public ParallelStreamProcessor getParallelConsumer(){
        org.apache.kafka.clients.consumer.Consumer<String, String> kafkaConsumer = getKafkaConsumer(); // (1)

        ParallelConsumerOptions<String, String> options = ParallelConsumerOptions.<String, String>builder()
                .ordering(KEY) // (2)
                .maxConcurrency(1000) // (3)
                .consumer(kafkaConsumer)
                .build();

        ParallelStreamProcessor<String, String> eosStreamProcessor =
                ParallelStreamProcessor.createEosStreamProcessor(options);

        eosStreamProcessor.subscribe(of(inputTopic)); // (4)

        return eosStreamProcessor;
    }

    org.apache.kafka.clients.consumer.Consumer<String, String> getKafkaConsumer() {
        Properties properties = new Properties();
        properties.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
        properties.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        properties.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
        properties.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
        return new KafkaConsumer<>(properties);
    }

    public void consume () {
        getParallelConsumer().poll(record ->
                log.info("Concurrently processing a record: {}", record));
    }

}
```

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'parallelConsumer' defined in file [.../ParallelConsumer.class]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.amway.magic.bonus.kafka.ParallelConsumer]: Constructor threw exception; nested exception is java.lang.NullPointerException
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1318) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1216) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:571) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:531) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:944) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:923) ~[spring-context-5.3.3.jar:5.3.3]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:588) ~[spring-context-5.3.3.jar:5.3.3]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144) ~[spring-boot-2.4.2.jar:2.4.2]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:767) [spring-boot-2.4.2.jar:2.4.2]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759) [spring-boot-2.4.2.jar:2.4.2]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:426) [spring-boot-2.4.2.jar:2.4.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:326) [spring-boot-2.4.2.jar:2.4.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1311) [spring-boot-2.4.2.jar:2.4.2]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1300) [spring-boot-2.4.2.jar:2.4.2]
	at com.amway.magic.bonus.KafkaPocApplication.main(KafkaPocApplication.java:10) [classes/:na]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [ParallelConsumer]: Constructor threw exception; nested exception is java.lang.NullPointerException
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:225) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:87) ~[spring-beans-5.3.3.jar:5.3.3]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1310) ~[spring-beans-5.3.3.jar:5.3.3]
	... 18 common frames omitted
Caused by: java.lang.NullPointerException: null
	at io.confluent.parallelconsumer.ParallelEoSStreamProcessor.checkAutoCommitIsDisabled(ParallelEoSStreamProcessor.java:323) ~[parallel-consumer-core-0.3.0.1.jar:na]
	at io.confluent.parallelconsumer.ParallelEoSStreamProcessor.<init>(ParallelEoSStreamProcessor.java:204) ~[parallel-consumer-core-0.3.0.1.jar:na]
	at io.confluent.parallelconsumer.ParallelStreamProcessor.createEosStreamProcessor(ParallelStreamProcessor.java:27) ~[parallel-consumer-core-0.3.0.1.jar:na]
	at com.amway.magic.bonus.kafka.ParallelConsumer.getParallelConsumer(ParallelConsumer.java:35) ~[classes/:na]
	at com.amway.magic.bonus.kafka.ParallelConsumer.consume(ParallelConsumer.java:52) ~[classes/:na]
	at com.amway.magic.bonus.kafka.ParallelConsumer.<init>(ParallelConsumer.java:22) ~[classes/:na]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_281]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_281]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_281]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_281]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:212) ~[spring-beans-5.3.3.jar:5.3.3]
	... 20 common frames omitted
```
I downloaded apache kafka from https://kafka.apache.org/downloads the current version is 2.13-2.7.0

@astubbs Yes, I got an exception after running my spring boot container 
```
Caused by: java.lang.UnsupportedClassVersionError: pl/tlinkowski/unij/service/collect/jdk8/Jdk8UnmodifiableListFactory has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
	at java.lang.ClassLoader.defineClass1(Native Method) ~[na:1.8.0_281]
	at java.lang.ClassLoader.defineClass(ClassLoader.java:756) ~[na:1.8.0_281]
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142) ~[na:1.8.0_281]
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468) ~[na:1.8.0_281]
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74) ~[na:1.8.0_281]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369) ~[na:1.8.0_281]
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363) ~[na:1.8.0_281]
	at java.security.AccessController.doPrivileged(Native Method) ~[na:1.8.0_281]
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362) ~[na:1.8.0_281]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[na:1.8.0_281]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355) ~[na:1.8.0_281]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[na:1.8.0_281]
	at java.lang.Class.forName0(Native Method) ~[na:1.8.0_281]
	at java.lang.Class.forName(Class.java:348) ~[na:1.8.0_281]
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:370) ~[na:1.8.0_281]
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404) ~[na:1.8.0_281]
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480) ~[na:1.8.0_281]
	at java.lang.Iterable.forEach(Iterable.java:74) ~[na:1.8.0_281]
	at pl.tlinkowski.unij.api.UniJLoader.load(UniJLoader.java:42) ~[pl.tlinkowski.unij.api-0.1.1.jar:na]
	at pl.tlinkowski.unij.api.UniJ.listFactory(UniJ.java:32) ~[pl.tlinkowski.unij.api-0.1.1.jar:na]
	at pl.tlinkowski.unij.api.UniLists.of(UniLists.java:60) ~[pl.tlinkowski.unij.api-0.1.1.jar:na]
	at ...ParallelConsumer.getParallelConsumer(ParallelConsumer.java:37) ~[classes/:na]
	at ..ParallelConsumer.consume(ParallelConsumer.java:53) ~[classes/:na]
	at ..ParallelConsumer.<init>(ParallelConsumer.java:22) ~[classes/:na]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_281]
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_281]
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_281]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423) ~[na:1.8.0_281]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:212) ~[spring-beans-5.3.3.jar:5.3.3]
	... 20 common frames omitted
```

--- a/.github/CODEOWNERS
+++ b/.github/CODEOWNERS
@@ -1,0 +1,1 @@
+*   astubbs

--- a/.github/workflows/maven.yml
+++ b/.github/workflows/maven.yml
@@ -32,7 +32,7 @@ jobs:
             name: "Newest AK version 2.7.0+ expected to fail"
 
     continue-on-error: ${{ matrix.experimental }}
-    name: "${{ matrix.name }} (AK: ${{ matrix.ak }}) experimental? ${{ matrix.experimental }} JDK: ${{ matrix.jdk }}"
+    name: "AK: ${{ matrix.ak }} JDK: ${{ matrix.jdk }}"
     runs-on: ubuntu-latest
 
     steps:

--- a/parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
+++ b/parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java
@@ -5,11 +5,18 @@ package io.confluent.parallelconsumer;
  */
 
 import io.confluent.csid.utils.WallClock;
-import lombok.*;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
-import org.apache.kafka.clients.consumer.*;
+import org.apache.kafka.clients.consumer.ConsumerRebalanceListener;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.MockConsumer;
 import org.apache.kafka.clients.consumer.internals.ConsumerCoordinator;
-import org.apache.kafka.clients.producer.*;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
 import org.apache.kafka.common.TopicPartition;
 import org.apache.kafka.common.utils.Time;
 import org.slf4j.MDC;
@@ -26,13 +33,12 @@ import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.regex.Pattern;
 
-import static io.confluent.parallelconsumer.ParallelEoSStreamProcessor.State.*;
 import static io.confluent.csid.utils.BackportUtils.isEmpty;
 import static io.confluent.csid.utils.BackportUtils.toSeconds;
 import static io.confluent.csid.utils.StringUtils.msg;
+import static io.confluent.parallelconsumer.ParallelEoSStreamProcessor.State.*;
 import static io.confluent.parallelconsumer.UserFunctions.carefullyRun;
 import static java.time.Duration.ofMillis;
-import static java.time.Duration.ofSeconds;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 import static java.util.concurrent.TimeUnit.SECONDS;
 
@@ -200,6 +206,7 @@ public class ParallelEoSStreamProcessor<K, V> implements ParallelStreamProcessor
 
         this.consumer = options.getConsumer();
 
+        checkGroupIdConfigured(consumer);
         checkNotSubscribed(consumer);
         checkAutoCommitIsDisabled(consumer);
 
@@ -223,12 +230,21 @@ public class ParallelEoSStreamProcessor<K, V> implements ParallelStreamProcessor
         }
     }
 
+    private void checkGroupIdConfigured(final org.apache.kafka.clients.consumer.Consumer<K, V> consumer) {
+        try {
+            consumer.groupMetadata();
+        } catch (RuntimeException e) {
+            throw new IllegalArgumentException("Error validating Consumer configuration - no group metadata - missing a " +
+                    "configured GroupId on your Consumer?", e);
+        }
+    }
+
     protected ThreadPoolExecutor setupWorkerPool(int poolSize) {
         ThreadFactory defaultFactory = Executors.defaultThreadFactory();
         ThreadFactory namingThreadFactory = r -> {
             Thread thread = defaultFactory.newThread(r);
             String name = thread.getName();
-            thread.setName("pc-"+name);
+            thread.setName("pc-" + name);
             return thread;
         };
         ThreadPoolExecutor.AbortPolicy rejectionHandler = new ThreadPoolExecutor.AbortPolicy();
@@ -333,12 +349,15 @@ public class ParallelEoSStreamProcessor<K, V> implements ParallelStreamProcessor
             coordinatorField.setAccessible(true);
             ConsumerCoordinator coordinator = (ConsumerCoordinator) coordinatorField.get(consumer); //IllegalAccessException
 
+            if (coordinator == null)
+                throw new IllegalStateException("Coordinator for Consumer is null - missing GroupId? Reflection broken?");
+
             Field autoCommitEnabledField = coordinator.getClass().getDeclaredField("autoCommitEnabled");
             autoCommitEnabledField.setAccessible(true);
             Boolean isAutoCommitEnabled = (Boolean) autoCommitEnabledField.get(coordinator);
 
             if (isAutoCommitEnabled)
-                throw new IllegalStateException("Consumer auto commit must be disabled, as commits are handled by the library.");
+                throw new IllegalArgumentException("Consumer auto commit must be disabled, as commits are handled by the library.");
         } else {
             // noop - probably MockConsumer being used in testing - which doesn't do auto commits
         }

--- a/parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessorTest.java
+++ b/parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessorTest.java
@@ -8,11 +8,10 @@ import io.confluent.csid.utils.LatchTestUtils;
 import io.confluent.parallelconsumer.ParallelConsumerOptions.CommitMode;
 import lombok.SneakyThrows;
 import lombok.extern.slf4j.Slf4j;
-import org.apache.kafka.clients.consumer.ConsumerGroupMetadata;
-import org.apache.kafka.clients.consumer.ConsumerRecord;
-import org.apache.kafka.clients.consumer.ConsumerRecords;
-import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.clients.consumer.*;
 import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serdes;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
@@ -23,9 +22,7 @@ import org.mockito.ArgumentMatchers;
 import org.mockito.Mockito;
 
 import java.time.Duration;
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
+import java.util.*;
 import java.util.concurrent.CountDownLatch;
 import java.util.function.Function;
 
@@ -718,7 +715,7 @@ public class ParallelEoSStreamProcessorTest extends ParallelEoSStreamProcessorTe
         if (commitMode.equals(PERIODIC_TRANSACTIONAL_PRODUCER)) {
             assertThatThrownBy(() -> parallelConsumer = initAsyncConsumer(optionsWithClients))
                     .isInstanceOf(IllegalArgumentException.class)
-                    .hasMessageContaining("Producer", "Transaction");
+                    .hasMessageContainingAll("Producer", "Transaction");
         } else {
             parallelConsumer = initAsyncConsumer(optionsWithClients);
             attachLoopCounter(parallelConsumer);
@@ -742,7 +739,7 @@ public class ParallelEoSStreamProcessorTest extends ParallelEoSStreamProcessorTe
     }
 
     @Test
-    void produceMessageFlowRequiresProducer() {
+    void optionsProduceMessageFlowRequiresProducer() {
         setupClients();
 
         var optionsWithClients = ParallelConsumerOptions.<String, String>builder()
@@ -752,8 +749,51 @@ public class ParallelEoSStreamProcessorTest extends ParallelEoSStreamProcessorTe
 
         assertThatThrownBy(() -> parallelConsumer = initAsyncConsumer(optionsWithClients))
                 .isInstanceOf(IllegalArgumentException.class)
-                .hasMessageContaining("Producer", "Transaction");
+                .hasMessageContainingAll("Producer", "Transaction");
     }
+
+
+    @Test
+    void optionsGroupIdRequiredAndAutoCommitDisabled() {
+        Properties properties = new Properties();
+        properties.setProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
+        Deserializer<String> deserializer = Serdes.String().deserializer();
+        var realConsumer = new KafkaConsumer<>(properties, deserializer, deserializer);
+
+        var optionsBuilder = ParallelConsumerOptions.<String, String>builder()
+                .consumer(realConsumer)
+                .commitMode(PERIODIC_CONSUMER_ASYNCHRONOUS);
+        var optionsWithClients = optionsBuilder
+                .build();
+
+        // fail
+        assertThatThrownBy(() -> parallelConsumer = initAsyncConsumer(optionsWithClients))
+                .as("Should error on missing group id")
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContainingAll("Consumer", "GroupId");
+
+        // add missing group id, now auto commit should fail
+        properties.setProperty(ConsumerConfig.GROUP_ID_CONFIG, "dummy-group");
+        optionsBuilder.consumer(new KafkaConsumer<>(properties, deserializer, deserializer));
+        assertThat(catchThrowable(() -> parallelConsumer = initAsyncConsumer(optionsBuilder.build())))
+                .as("Should error on auto commit enabled by default")
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContainingAll("auto", "commit", "disabled");
+
+        // fail auto commit disabled
+        properties.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "true");
+        optionsBuilder.consumer(new KafkaConsumer<>(properties, deserializer, deserializer));
+        assertThat(catchThrowable(() -> parallelConsumer = initAsyncConsumer(optionsBuilder.build())))
+                .as("Should error on auto commit enabled")
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContainingAll("auto", "commit", "disabled");
+
+        // set missing auto commit
+        properties.setProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+        optionsBuilder.consumer(new KafkaConsumer<>(properties, deserializer, deserializer));
+        assertThatNoException().isThrownBy(() -> parallelConsumer = initAsyncConsumer(optionsBuilder.build()));
+    }
+
 
     @Test
     void cantUseProduceFlowWithWrongOptions() throws InterruptedException {
@@ -776,7 +816,7 @@ public class ParallelEoSStreamProcessorTest extends ParallelEoSStreamProcessorTe
         assertThatThrownBy(() -> parallel.pollAndProduce((record) ->
                 new ProducerRecord<>(INPUT_TOPIC, "hi there")))
                 .isInstanceOf(IllegalArgumentException.class)
-                .hasMessageContaining("Producer", "Transaction");
+                .hasMessageContainingAll("Producer", "options");
     }
 
     @ParameterizedTest()

